File: chapter3/section3.3/range.py
"""
ID: riff.sc1
LANG: PYTHON3
TASK: range
"""

# This solution is too slow, but I believe it's correct.

def main():
  with open('range.in') as fin:
    farm_size = int(fin.readline())
    farm = []
    for _ in range(farm_size):
      farm.append(fin.readline().strip())

  # squares_of_size[n] is a 2D boolean array. squares_of_size[n][i][j] is True
  # iff the n by n square whose top-left corner is (i, j) is good for grazing.
  squares_of_size = {}
  size_two_squares = [[True for _ in range(farm_size - 1)] 
                            for _ in range(farm_size - 1)]
  squares_of_size[2] = size_two_squares

  for i in range(farm_size):
    for j in range(farm_size):
      if farm[i][j] == '0':
        invalidate(size_two_squares, i, j)

  for size in range(3, farm_size + 1):  # 3 .. farm_size inclusive
    bigger_squares = [[True for _ in range(farm_size - size + 1)] 
                            for _ in range(farm_size - size + 1)]
    squares_of_size[size] = bigger_squares
    smaller_squares = squares_of_size[size - 1]

    # Build 3x3 squares from 2x2 squares, etc.
    for i in range(farm_size - size + 2):  # farm_size - (size - 1) + 1
      for j in range(farm_size - size + 2):
        if not smaller_squares[i][j]:
          invalidate(bigger_squares, i, j)

  with open('range.out', 'w') as fout:
    for size in range(2, farm_size + 1):  # 2 .. farm_size inclusive
      num_good_squares = count_good_squares(squares_of_size[size])
      if num_good_squares > 0:
        fout.write(f'{size} {num_good_squares}\n')

def count_good_squares(squares):
  answer = 0
  for row in squares:
    answer += row.count(True)
  return answer

def invalidate(squares, row, col):
  size = len(squares)
  for i in range(row - 1, row + 1):  # (row - 1) .. row inclusive
    for j in range(col - 1, col + 1):
      if 0 <= i < size and 0 <= j < size:
        squares[i][j] = False

from time import time
start = time()
main()
print(f'{time() - start}')

File: chapter3/section3.3/range.cpp
/*
ID: riff.sc1
LANG: C++
TASK: range
*/
#include <fstream>
#include <string>

std::ifstream fin("range.in");
std::ofstream fout("range.out");
const int kMaxSize = 250;

// farm[n][i][j] indicates whether the n x n square with (i, j) as its top-left
// corner is a good square for grazing.
// Due to space constraints, we only save two maps at a time, rotating the first 
// index. 1x1 squares go into farm[0], 2x2 goes into farm[1], 3x3 goes into
// farm[0], 4x4 goes into farm[1], etc.
bool farm[2][kMaxSize][kMaxSize];

int main() {
  int farm_size;
  fin >> farm_size;

  // Process input.
  std::string farm_data[farm_size];
  for (int i = 0; i < farm_size; ++i) {
    fin >> farm_data[i];
  }

  for (int i = 0; i < farm_size; ++i) {
    for (int j = 0; j < farm_size; ++j) {
      if (farm_data[i][j] == '1') {
        farm[0][i][j] = true;
      } else {
        farm[0][i][j] = false;
      }
    }
  }

  int small = 0;
  int large = 1;
  for (int size = 2; size <= farm_size; ++size) {
    // Build 2x2 squares from 1x1 squares, etc.
    int num_good_squares = 0;
    for (int i = 0; i < farm_size - size + 1; ++i) {
      for (int j = 0; j < farm_size - size + 1; ++j) {
        if (farm[small][i][j] && farm[small][i+1][j] && 
            farm[small][i][j+1] && farm[small][i+1][j+1]) {
          farm[large][i][j] = true;
          ++num_good_squares;
        } else {
          farm[large][i][j] = false;
        }
      }
    }
    if (num_good_squares > 0) {
      fout << size << ' ' << num_good_squares << std::endl;
    }

    // Swap the two arrays.
    small = 1 - small;
    large = 1 - large;
  }

  return 0;
}